[8.x] n-api: backport changes from master

gabrielschulhof opened the pull request:
The following commits (or relevant portions thereof) are included:

    * c123467 timers: allow Immediates to be unrefed (the portion fixing
      the uv_test_loop test)
    * a555397 n-api: add methods to open/close callback scope
    * 47f664e n-api: wrap control flow macro in do/while
    * 1286923 n-api: implement wrapping using private properties
    * 316b5ef n-api: throw RangeError napi_create_typedarray()
    * 8938c4c n-api: expose n-api version in process.versions
    * 91c1ccd n-api: throw RangeError in napi_create_dataview() with
      invalid range
    * 94e2951 n-api: fix memory leak in napi_async_destroy()
    * dc389bf n-api: use nullptr instead of NULL in node_api.cc
    * a4a9a3d src: add napi_handle_scope_mismatch to msg list

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

gabrielschulhof commented:
Fixes https://github.com/nodejs/abi-stable-node/issues/297

gabrielschulhof commented:
This also fails on an unrelated test: `parallel/test-benchmark-misc`.

MylesBorins commented:
@gabrielschulhof we have generally backported every commit individualls... that makes it easier to audit in the future and stop tooling from breaking

@nodejs/lts thoughts on landing a single commit comprising of many patches?

gibfahn commented:
>@nodejs/lts thoughts on landing a single commit comprising of many patches?

I'd rather backport each commit individually, although I'm +1 on it being in one PR (i.e. this one). @gabrielschulhof if you could split this into the individual commits that would be good. You don't need to change any commit metadata, so it should just be a case of cherry-picking and fixing merge conflicts.

gabrielschulhof commented:
Can do.

On Thu, Mar 22, 2018 at 11:45 AM, Gibson Fahnestock <
notifications@github.com> wrote:

> @nodejs/lts <https://github.com/orgs/nodejs/teams/lts> thoughts on
> landing a single commit comprising of many patches?
>
> I'd rather backport each commit individually, although I'm +1 on it being
> in one PR (i.e. this one). @gabrielschulhof
> <https://github.com/gabrielschulhof> if you could split this into the
> individual commits that would be good. You don't need to change any commit
> metadata, so it should just be a case of cherry-picking and fixing merge
> conflicts.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/19265#issuecomment-375354085>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA7k0UYxc8Ju-XtmJ8yYplmjJ0YArQK-ks5tg8c0gaJpZM4Sksat>
> .
>

gabrielschulhof commented:
@gibfahn there are some commits where I've intentionally not applied parts of it even though they would've applied cleanly, because those parts were not related to N-API.

gabrielschulhof commented:
@gibfahn @MylesBorins I have now replaced the single commit with the sometimes-partially cherry-picked individual commits.

gabrielschulhof commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/13825/

gabrielschulhof commented:
The same unrelated failure has reoccurred.

gibfahn commented:
>@gibfahn there are some commits where I've intentionally not applied parts of it even though they would've applied cleanly, because those parts were not related to N-API.

So if you think those changes should be applied to 8.x then you should leave them in, we try to backport whole commits where possible, otherwise understanding what's been backported becomes even harder.

If they don't need to go back to 8.x at all, then leaving them out is fine.

Sorry to keep complaining about this PR, I really appreciate you doing this work!

gabrielschulhof commented:
@gibfahn NP. We have to do things right.

I'll do the cherry-picking again, being more inclusive. Any decisions I make I will document in a subsequent comment here.

gabrielschulhof commented:
@MylesBorins @gibfahn the list of commits from master:

⁠ | ⁠ | ⁠ | ⁠ | ⁠ | ⁠
--- | --- | --- | --- | --- | ---
6572f63afb | d4b3fcba71 | c6c957d3cc | be470943fb | c1695d8bad | f24d0ec201
6a5a9ad62d | d20f1f004c | 5fb6f7f22f | cd7d7b15c1 | a03c90b661 | 33e63fe64f
48b5c11b16 | 35c7238bb7 | d83f8303fd | 449d1c8ea4 | c698017a2f | a29089d7c8
9cb96ac828 | 6e1c25c456 | 463d1a490f | caee112e52 | 755e07cb73 | a1bab826c4
2bead4ba9e | 1729af2ce9 | fe442f6b5f | d3569b623c | 5ecd2e15b8 | 6101bd209e
d379db31fd | a555397185 | c2b9048ac8 | 629a4476f0 | 47f664ee88 | 1286923cc0
0c16b18043 | 6e312c5cc7 | 6c1906ab3e | c1234673bb | f054855bbf | c90185c15e
316b5efd6b | f75bc2c1a5 | 3e062762e4 | 8938c4c22d | 91c1ccd84f | 8a86d9c1cf
ef49f55e93 | | | | |

These are the places where I made a judgement call:

* c1234673bbba1ac6c8425dffb2604ccf647bbfcf timers: allow Immediates to be unrefed
    I did not backport the `immediate.ref()` and `immediate.unref()`, but only the modification to `test/addons-napi/test_uv_loop/test_uv_loop.cc`
* d3569b623ccd593c9ef62fcaf0aba2711dc7fbfa doc: remove &#42;&#42;Note:&#42;&#42; tags
    This touched N-API, so I grabbed the whole thing, dropping chunks which do not apply, and manually applying the fix in places where there are such tags that were not on master
* caee112e52b64f4bc1118c4a5fa5ad7b4211efea test: remove assert.doesNotThrow()
  I removed the files that were "deleted by us", checked out "our" version of those that conflicted, and removed `assert.doesNotThrow()` by hand
* 463d1a490f2ed4679f8805fa57084b4445ea40d6 test,benchmark,doc: enable dot-notation rule
    I removed the files that were "deleted by us", checked out "our" version of those that conflicted, and moved to dot-notation by hand. I ran the new linter to make sure everything was covered.
* 6e1c25c45672b70f4b6c6c8af56d9c0762bfae04 errors: update all internal errors
    I removed the unused errors but I did not backport the updated definition of `E()` which accepts an additional parameter.
* 9cb96ac82808254e3322119eb1e6b2a7b847741e doc: remove extraneous "for example" text
    Did my best to merge this, since its documentation-only.
* a29089d7c866955616c0e363843017e9b9b2a736 doc: add new documentation lint rule
    I applied as much of the commit as possible, checking out "our" version of conflicting files, and manually wrapping lines at 80 characters to conform to the linting rule.
* 35c7238bb793ee6c6aa9f5f4f3577e94ea2b9837 doc: replace to Node.js
    I applied as much of the commit as possible, checking out "our" version of conflicting files, and using `git grep -niH 'node' -- doc/api doc/guides tools/icu/README.md | grep -v 'Node[.]js' | grep -i --color=always node | less -R` to find places that should read "Node.js" instead of "node". I found one place where it's better to replace "node" with "&#96;node&#96;" rather than "Node.js", because the text refers to the name of the binary.

gabrielschulhof commented:
I'll close and open another because this needs to go on top of v8.x-staging.

addaleax commented:
@gabrielschulhof Not sure if that’s what you mean by your last comment, but using the same button that you’d use to change a PR title, you can also change its base branch :)

gabrielschulhof commented:
@addaleax Oooh! Thanks!

gabrielschulhof commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/13911/

gabrielschulhof commented:
I've updated the list of commits above because I had missed some in my previous listing. No additional judgement calls though.

CI: https://ci.nodejs.org/job/node-test-pull-request/13913/

gabrielschulhof commented:
CI after rebase: https://ci.nodejs.org/job/node-test-pull-request/13959/

gibfahn commented:
New CI as the previous one failed on ARM centos.

CI: https://ci.nodejs.org/job/node-test-commit/17393/

gabrielschulhof commented:
... and I removed some broad and shallow commits to improve reviewability:

-doc: fix various nits
-doc: replace to Node.js
-doc: add new documentation lint rule
-doc: remove extraneous "for example" text
-test,benchmark,doc: enable dot-notation rule
-test: remove assert.doesNotThrow()
-doc: remove **Note:** tags
-src: remove duplicate words in comments

gabrielschulhof commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/14106/

gabrielschulhof commented:
Added another commit, so here's a new CI: https://ci.nodejs.org/job/node-test-pull-request/14132/

gabrielschulhof commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/14168/

gabrielschulhof commented:
I removed some broad-and-shallow commits to focus the PR. New CI: https://ci.nodejs.org/job/node-test-pull-request/14187/

MylesBorins commented:
216bfd7 is Semver-Minor. We can land this entire PR and wait for an upcoming minor release to land that specific commit (rather than blocking the entire thing)

gabrielschulhof commented:
@MylesBorins I dropped that commit. New CI: https://ci.nodejs.org/job/node-test-pull-request/14228/

gabrielschulhof commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/14318/

gabrielschulhof commented:
@nodejs/collaborators may I have some reviews of this PR, please?

mhdawson commented:
New CI since link is stale: https://ci.nodejs.org/job/node-test-pull-request/14532/

mhdawson commented:
To make reviewing easier this is the high level summary:

**DOCS only**
doc/api/errors.md - pulls in desc of new N-API errors
doc/api/n-api.md - updates N-API docs

**Non N-API specific files**
lib/internal/errors.js - addes N-API errors
src/env.h - add N-API private symbols
src/node.cc - adds N-API version to env
src/node_internals.h - FatalException moved here 
src/node_version.h - updates NAPI_VERSION

**Non N-API test files**
test/parallel/test-process-versions.js - adds napi version to test

**N-API specific files**
src/node_api.cc 
src/node_api.h
src/node_api_types.h
~41 N-API test files

Change to non N-API parts are small and low risk
Changes to N-API parts bring it in line with master and look good.

gabrielschulhof commented:
@mhdawson thanks for reviewing!

MylesBorins commented:
It seems like e15f57745d  was only a partial backport of a test from a semver-minor commit... this will make it impossible for us to audit that semver-minor commit in the future if we wanted to land it.

Were there any other commits / prs that were partially backported?

gabrielschulhof commented:
@MylesBorins 2838f9b150 doesn't modify test/parallel/test-console-assign-undefined.js whereas the original does.

I used
```bash
git log --oneline 3225601...07a6770 | \
while read id message; do \
  echo $id; \
  diff -u \
    <(git show --stat=99 $id | grep -vE '^commit|files changed' | grep '|') \
    <(git log --oneline master | grep "$message" | awk '{print $1;}' | xargs \
        git show --stat=99 | grep -vE '^commit|files changed' | grep '|'); \
done```

to check whether the files touched by the backported commits differ from the files touched by the commits on master.

MylesBorins commented:
@gabrielf did that script identify `e15f577` as well? If so this satisfied me!

Perhaps we should consider putting this into node-core-utils to help us with reviewing backports /cc @nodejs/automation

gabrielf commented:
@MylesBorins you are asking the wrong Gabriel...

gabrielschulhof commented:
@gabrielf sorry!

@MylesBorins it did. This was the output:

```
07a6770614
8b3ef4660a
92f699e021
367113f5d7
f0ba2c6ceb
24b8bb6708
3a6b7e610d
9949d55ae9
f29d8e0e8d
7c6fa183cb
584fadc605
4c1181dc02
faf94b1c49
df63adf7aa
b26410e86f
1abb168838
cb3f90a1a9
6129ff4b99
87d0fd8212
58688d97dc
2838f9b150
--- /dev/fd/63	2018-05-03 15:57:59.614213570 -0400
+++ /dev/fd/62	2018-05-03 15:57:59.614213570 -0400
@@ -1 +1,2 @@
- test/addons-napi/test_typedarray/test.js | 4 ++--
+ test/addons-napi/test_typedarray/test.js       | 4 ++--
+ test/parallel/test-console-assign-undefined.js | 3 +--
5c0983e5a2
4d43607474
--- /dev/fd/63	2018-05-03 15:58:00.621202048 -0400
+++ /dev/fd/62	2018-05-03 15:58:00.621202048 -0400
@@ -1 +1,2 @@
  doc/api/n-api.md | 2 +-
+ doc/api/n-api.md | 2 +-
9244e1d234
927fc0b19f
9729278007
7ed1dfef28
969a520990
d89f5937eb
bebcdfe382
af655f586c
5b1b74c5a5
e15f57745d
--- /dev/fd/63	2018-05-03 15:58:05.856142148 -0400
+++ /dev/fd/62	2018-05-03 15:58:05.856142148 -0400
@@ -1 +1,10 @@
- test/addons-napi/test_uv_loop/test_uv_loop.cc | 9 +++++++++
+ doc/api/timers.md                                   |  32 +++++++++
+ lib/timers.js                                       | 143 +++++++++++++++++++++++----------------
+ src/env-inl.h                                       |  40 +++++++++--
+ src/env.cc                                          |  36 +++++-----
+ src/env.h                                           |  17 ++++-
+ src/node_perf.cc                                    |  18 ++---
+ src/timer_wrap.cc                                   |  12 ++--
+ test/addons-napi/test_uv_loop/test_uv_loop.cc       |   9 +++
+ test/parallel/test-timers-immediate-unref-simple.js |   7 ++
+ test/parallel/test-timers-immediate-unref.js        |  37 ++++++++++
84e0a03727
8cfa87832d
ca10fda064
51513fdf9e
02ae3295d5
853b4d593c
48be8a4793
79ecc2c586
ad8c079af7
a744535f99
2e100c82be
077e1870ae
```

I didn't mention 4d43607474 because there are two commits with the same message on master.
